subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/DefaultConfiguration.java

```diff
@@ -222,6 +223,12 @@ public class DefaultConfiguration extends AbstractFileCollection implements Conf
     private boolean canBeConsumed = true;
     private boolean canBeResolved = true;
     private boolean canBeDeclaredAgainst = true;
+    private boolean consumptionDeprecated = false;
```

Comment:
Instead of saving all of these as individual fields, what do you think of saving a single `ConfigurationRole` to encapsulate all this state? This way, we'd also be able to get rid of `ConfigurationRole#forConfiguration`, which seems like a pretty odd method

Reply:
So, keep a `roleAtCreation` and a `actualRole`?  I suppose the actual role would be another, mutable implementation of `ConfigurationRole`, with a getter that returns an immutable copy?

I'm slightly against it, since I think it's more complex than the one extra factory method, which I don't find that weird.

Reply:
Maybe I just think it's odd that forConfiguration lives inside ConfigurationRole, it seems like that logic would fit better in DefaultConfiguration? It's nbd tho

Reply:
Maybe `fromConfiguration` would be better?  Or maybe I should just drop it and inline the logic to build a second role for usage comparison.

Reply:
I removed the forConfiguration method.